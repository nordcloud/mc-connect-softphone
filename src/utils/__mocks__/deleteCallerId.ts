export const deleteCallerId = jest.fn().mockResolvedValue('dummyCallerId');
