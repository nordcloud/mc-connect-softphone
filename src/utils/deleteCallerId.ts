import { DynamoDBClient } from '@aws-sdk/client-dynamodb';
import { DeleteCommand, DynamoDBDocumentClient } from '@aws-sdk/lib-dynamodb';

import { CALLER_ID_TABLE } from '../consts';

const ddbClient = new DynamoDBClient({});
const ddbDocClient = DynamoDBDocumentClient.from(ddbClient);

export async function deleteCallerId(username: string) {
  const deleteCommand = new DeleteCommand({
    TableName: CALLER_ID_TABLE,
    Key: { username },
    ReturnValues: 'ALL_OLD',
  });

  const { Attributes } = await ddbDocClient.send(deleteCommand);

  return Attributes?.callerId;
}
