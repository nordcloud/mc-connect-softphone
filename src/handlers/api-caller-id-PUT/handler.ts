import { APIGatewayProxyHandlerV2 } from 'aws-lambda';
import { StatusCodes } from 'http-status-codes';

import { getUser } from '../../utils/getUser';
import { putCallerId } from './putCallerId';

export const handler: APIGatewayProxyHandlerV2 = async (event) => {
  const user = await getUser(event);

  if (!user) {
    return { statusCode: StatusCodes.UNAUTHORIZED };
  }

  const callerId = event.body || '';
  const { expiresTimestampSeconds } = await putCallerId({ user, callerId });

  return {
    statusCode: StatusCodes.OK,
    body: JSON.stringify({ expiresTimestampSeconds }),
  };
};
