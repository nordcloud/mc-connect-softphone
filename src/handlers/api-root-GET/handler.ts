import { APIGatewayProxyHandlerV2 } from 'aws-lambda';

import { AWS_CONNECT_URL } from '../../consts';
import { deleteCallerId } from '../../utils/deleteCallerId';
import { getCustomers } from '../../utils/getCustomers';
import { getUser } from '../../utils/getUser';
import { loginRedirect } from '../../utils/loginRedirect';
import { renderHTML } from '../../utils/renderHTML';

export const handler: APIGatewayProxyHandlerV2 = async (event) => {
  const user = await getUser(event);

  if (!user) {
    return loginRedirect();
  }

  await deleteCallerId(user.email);

  const customers = await getCustomers();

  return renderHTML('homepage', {
    AWS_CONNECT_URL,
    user,
    customers,
  });
};
