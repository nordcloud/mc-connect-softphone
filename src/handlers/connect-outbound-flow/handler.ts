import { ConnectContactFlowHandler } from 'aws-lambda';

import { getCallerId } from '../../utils/getCallerId';

export const handler: ConnectContactFlowHandler = async (event) => {
  const username = event.Details.Parameters.Username;

  if (!username) {
    throw new Error('Missing username');
  }

  const callerId = await getCallerId(username);

  return {
    caller_id: callerId || null,
  };
};
