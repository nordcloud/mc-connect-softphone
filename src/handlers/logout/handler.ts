import { APIGatewayProxyHandlerV2 } from 'aws-lambda';
import cookie from 'cookie';
import { StatusCodes } from 'http-status-codes';

import { ID_TOKEN_COOKIE } from '../../consts';
import { getCookie } from '../../utils/getCookie';
import { getStageConsts } from '../../utils/getStageConsts';

const { OAUTH_BASE_URL } = getStageConsts();

export const handler: APIGatewayProxyHandlerV2 = async (event) => {
  const { API_URL } = process.env;

  if (!API_URL) {
    throw new Error('Missing API_URL');
  }

  const idToken = getCookie(event, ID_TOKEN_COOKIE);
  let location = `${OAUTH_BASE_URL}/logout?id_token_hint=${idToken}&post_logout_redirect_uri=${API_URL}`;

  if (!idToken) {
    console.log('Missing idToken cookie');
    location = '/';
  }

  console.log(`Redirecting to ${location}`);

  return {
    statusCode: StatusCodes.MOVED_TEMPORARILY,
    headers: { location },
    cookies: [cookie.serialize(ID_TOKEN_COOKIE, '')],
  };
};
