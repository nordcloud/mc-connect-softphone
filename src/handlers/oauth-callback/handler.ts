import { APIGatewayProxyHandlerV2 } from 'aws-lambda';
import cookie from 'cookie';
import { StatusCodes } from 'http-status-codes';

import { ID_TOKEN_COOKIE } from '../../consts';
import { getStage } from '../../utils/getStage';
import { fetchIdToken } from './fetchIdToken';

export const handler: APIGatewayProxyHandlerV2 = async (event) => {
  const responseCookies = [];
  const idTokenData = await fetchIdToken(event);

  if (idTokenData) {
    responseCookies.push(
      cookie.serialize(ID_TOKEN_COOKIE, idTokenData.id_token, {
        secure: getStage() !== 'local',
        httpOnly: true,
        expires: new Date(Date.now() + idTokenData.expires_in * 1000),
      })
    );
  }

  console.log('Redirecting to "/"');

  return {
    statusCode: StatusCodes.MOVED_TEMPORARILY,
    headers: { location: '/' },
    cookies: responseCookies,
  };
};
