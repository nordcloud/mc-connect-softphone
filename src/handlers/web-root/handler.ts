import { APIGatewayProxyHandlerV2 } from 'aws-lambda';

import { AWS_CONNECT_URL } from '../../consts';
import { getCustomers } from '../../utils/getCustomers';
import { getUser } from '../../utils/getUser';
import { loginRedirect } from '../../utils/loginRedirect';
import { renderHTML } from '../../utils/renderHTML';

export const handler: APIGatewayProxyHandlerV2 = async (event) => {
  const user = await getUser(event);

  if (!user) {
    return loginRedirect();
  }

  return renderHTML('homepage', {
    AWS_CONNECT_URL,
    user,
    customers: await getCustomers(event),
  });
};
