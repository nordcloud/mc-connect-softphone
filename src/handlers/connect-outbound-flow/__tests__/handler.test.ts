import { deleteCallerId } from '../../../utils/deleteCallerId';
import { handler } from '../handler';

jest.mock('../../../utils/deleteCallerId');

const args = [
  {
    Details: {
      Parameters: {
        Username: 'dummyUsername',
      },
    },
  },
] as unknown as Parameters<typeof handler>;

describe('connect-outbound-flow-handler', () => {
  it('calls deleteCallerId with correct args', async () => {
    await handler(...args);
    expect(jest.mocked(deleteCallerId).mock.calls).toMatchSnapshot();
  });

  it('resolves with a correct value', () => {
    return expect(handler(...args)).resolves.toMatchSnapshot();
  });

  it('resolves with a correct value when callerId returns a falsy value', () => {
    jest.mocked(deleteCallerId).mockResolvedValueOnce(undefined);
    return expect(handler(...args)).resolves.toMatchSnapshot();
  });

  it('rejects when Username is missing from event parameters', () => {
    const argsClone = structuredClone(args);
    delete argsClone[0].Details.Parameters.Username;
    return expect(handler(...argsClone)).rejects.toMatchSnapshot();
  });
});
