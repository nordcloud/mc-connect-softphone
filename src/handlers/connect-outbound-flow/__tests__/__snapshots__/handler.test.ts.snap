// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`connect-outbound-flow-handler calls getCallerId with correct args 1`] = `
[
  [
    "dummyUsername",
  ],
]
`;

exports[`connect-outbound-flow-handler rejects when Username is missing from event parameters 1`] = `[Error: Missing username]`;

exports[`connect-outbound-flow-handler resolves with a correct value 1`] = `
{
  "caller_id": "dummyCallerId",
}
`;

exports[`connect-outbound-flow-handler resolves with a correct value when callerId returns a falsy value 1`] = `
{
  "caller_id": null,
}
`;
